enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,327 @@
+combinations of same. At a minimum, we will explicitly test every legacy config
+option listed in the user documentation.
+
+**Risk** There may be additional configuration required for CCM compared to the
+legacy cloud provider in certain cloud configurations.
+
+**Mitigation** We currently test multiple cloud configurations. We should test
+an upgraded cluster on each of these cloud configurations.
+
+## Design Details
+
+### Configuration transformation for OpenStack CCM
+
+The configuration of OpenStack CCM is a superset of the legacy cloud provider
+configuration, with four exceptions. Details on how we will address each of
+these are provided below.
+
+<dl>
+<dt>secret-name</dt>
+<dt>secret-namespace</dt>
+<dt>kubeconfig-path</dt>
+<dd>
+  These are no longer relevant or supported in CCM. They define the cloud
+  credentials that are used by the legacy cloud provider. CCO will automatically
+  replace these with static config pointing to the cluster-managed cloud
+  credentials.
+  <br>
+  It is possible that the secret contains more information that the
+  cloud credentials populated by the installer. If so, the updated
+  configuration will no longer include these manual edits. This is
+  acceptable since they've done a dirty thing and deserve pain.
+</dd>
+<dt>node-volume-attach-limit</dt>
+<dd>
+  Along with all block storage options, this is now handled by Cinder CSI.
+  Unlike other block storage options, this is the only one that has been removed
+  and therefore the only one that will cause parse issues if included in
+  configuration. We will drop all block storage options when transforming the
+  user-provided configuration.
+</dd>
+</dl>
+
+We will always add `use-clouds=true` and `cloud="openstack"` in the generated
+`kube-cloud-config-external`.  This will refer to the cluster-managed cloud
+credentials. If the user specifies these values in the user-managed ConfigMap,
+they will be ignored and overridden.
+
+If the user specifies configuration that would not parse after transformation,
+we will mark CCO as degraded, and will not update the transformed configuration.
+
+### Open Questions [optional]
+
+1. After the upgrade to 4.11, we should give the user a warning if their
+   configuration is being transformed. The user should get a warning if they
+   are using static configuration such as `secret-name` and `secret-namespace`
+   and should be advised to remove this configuration. They should also see
+   warnings if they have any block-storage-related config options defined as
+   these are now handled by cinder-csi.
+
+2. Something needs to populate
+   `openshift-config-managed/kube-cloud-config-external` before CCM can run. In
+   a running cluster CCO would do this. However, CCM starts _before_ CCO which
+   presents a bootstrapping problem on initial installation. How should we
+   resolve this? Should we let the installer do this seeding or something else?
```

Comment:
Change config format which installer populates sounds adequate, i think. For now it might be done based on featuregate information, after OpenStack ccm GA it might become default behaviour.
But this will require installer changes then. As far as i'm aware, installer does not have mechanism for handling feature gates at this point. https://github.com/openshift/installer/pull/4640/files - there were some attempts to introduce feature gates handling there, but with no success yet.